src/utils/mod.rs

```diff
@@ -246,30 +247,54 @@ async fn download_file_(
     // Download the file
 
     // Keep the curl env var around for a bit
-    let use_curl_backend = process
-        .var_os("RUSTUP_USE_CURL")
-        .is_some_and(|it| it != "0");
-    let use_rustls = process
-        .var_os("RUSTUP_USE_RUSTLS")
-        .is_none_or(|it| it != "0");
-    let (backend, notification) = if use_curl_backend {
-        (Backend::Curl, Notification::UsingCurl)
-    } else {
-        let tls_backend = if use_rustls {
-            TlsBackend::Rustls
-        } else {
-            #[cfg(feature = "reqwest-native-tls")]
-            {
-                TlsBackend::NativeTls
+    let use_curl_backend = process.var_os("RUSTUP_USE_CURL").map(|it| it != "0");
+    let use_rustls = process.var_os("RUSTUP_USE_RUSTLS").map(|it| it != "0");
+
+    let backend = match (use_curl_backend, use_rustls) {
+        // If environment specifies a backend that's unavailable, error out
+        #[cfg(not(feature = "reqwest-rustls-tls"))]
+        (_, Some(true)) => {
+            return Err(anyhow!(
+                "RUSTUP_USE_RUSTLS is set, but reqwest-rustls-tls is not enabled"
+            ));
+        }
+        #[cfg(not(feature = "reqwest-native-tls"))]
+        (_, Some(false)) => {
+            return Err(anyhow!(
+                "RUSTUP_USE_RUSTLS is set to false, but reqwest-native-tls is not enabled"
+            ));
+        }
+        #[cfg(not(feature = "curl-backend"))]
+        (Some(true), _) => {
+            return Err(anyhow!(
+                "RUSTUP_USE_CURL is set, but curl-backend is not enabled"
+            ));
+        }
+        #[cfg(feature = "curl-backend")]
+        (Some(true), None) => Backend::Curl,
+        #[cfg(feature = "reqwest-native-tls")]
+        (_, Some(false)) => {
+            if use_curl_backend == Some(true) {
+                info!("RUSTUP_USE_CURL is set and RUSTUP_USE_RUSTLS is set to off, using reqwest with native-tls");
             }
-            #[cfg(not(feature = "reqwest-native-tls"))]
-            {
-                TlsBackend::Rustls
+            Backend::Reqwest(TlsBackend::NativeTls)
+        }
+        #[cfg(feature = "reqwest-rustls-tls")]
+        (_, _) => {
```

Comment:
The long match here is still a bit difficult to wrap my head around. I admit that the actual logic is already complex here, but I think you can at least be more explicit about `_` cases... For instance, can this be written as something like `(_, Some(true) | None)`?

Reply:
Maybe using a big match statement isn't ideal for this? The two values are almost entirely independent, except for when both are set.

Reply:
Done. IMO this kind of defeats the point -- the compiler will make sure that the arms are exhaustive so it feels redundant to restate all the variants covered by this arm.

Reply:
> Maybe using a big match statement isn't ideal for this? The two values are almost entirely independent, except for when both are set.

Hmm, that's fair, let me try another approach.

Reply:
Reworked this. Also did a bit more work to make sure that the `match` expressions are exhaustive in most combinations of the Cargo features.